### ISSUE TITLE
Update FindClusters to call Leiden in rigraph

### ISSUE THREAD

**Issue by TomKellyGenetics** [satijalab/seurat#3685]
The Leiden algorithm has been merged in to the development version of the R "igraph" package. This has considerably better performance than calling Leiden with reticulate and could remove the need for Python dependencies.

This change has been tested in the R "leiden" package and gives the same results for most test cases (although igraph::cluster_leiden does not support passing the random seed you can do it with `set.seed()` in R). The "leiden" R package supports S3 methods for sparse dgCMatrix class which Seurat::Graph inherits.

See here for details: https://github.com/TomKellyGenetics/leiden/issues/1#issuecomment-722157607

However, some changes will be needed for Seurat::FindClusters to benefit from improved performance. Namely, the "RBConfigurationVertexPartition" and directed graphs are not supported but "ModularityVertexPartition" on undirected  graphs is supported. This would require changing the default partition function to Seurat.

As noted in the [leidenalg documentation](https://leidenalg.readthedocs.io/en/latest/reference.html?highlight=rbc#leidenalg.RBConfigurationVertexPartition) these functions give the same results for `resolution = 1`.
> Note that this is the same as ModularityVertexPartition when setting 𝛾=1 and normalising by 2𝑚, or 𝑚 for directed graphs.

Weights are still supported and `igraph::cluster_leiden(graph, objective_function = "modularity)` supports a _resolution_ parameter. This would maintain most of the functions of `RunLeiden`.

**TL;DR: Benefits**

* Soft dependency on leidenalg with reticulate removes need for Python install 

* Much faster computation by calling C and C++ functions from R

* Only minor changes to Seurat needed to support this

**Caveats**

* Cannot release on CRAN due to remote dependencies (until r-igraph is updated).

* May change default behaviour of `FindClusters`

In the meantime leiden 0.3.4 (released recently) automatically configures conda for leidenalg and umap-learn upon loading if Python and conda is available.

**Comment by denvercal1234GitHub** [comment]
Hi @TomKellyGenetics - Thank you for the post. I am new to this, and wonder did you mean that it is now easier to use Leiden algorithm option in FindCluster() function? Because I thought setting algorithm=4 would use Leiden algorithm already?

**Comment by TomKellyGenetics** [comment]
It does already do that yes, I actually contributed [that pull request](https://github.com/satijalab/seurat/pull/1060) to add Leiden to `Seurat::FindClusters()` and maintain the CRAN `leiden` package. Currently this calls "leidenalg" in python using `reticulate`. What's proposed here is changes to backend to avoid this. The changes visible to users would be minimal but it should improve performance considerably. 

Changes to the C igraph core [to support leiden](https://github.com/igraph/igraph/pull/1259) and changes to the R igraph package [to call it in R](https://github.com/igraph/rigraph/issues/346) have been merged. However, it is not yet available on CRAN. When "igraph" version 1.2.7.0 is release on CRAN we are able to support it in "leiden" and "Seurat".
 
The new version of "igraph" is [available](https://github.com/igraph/rigraph/tree/dev) for testing on GitHub and is supported in the [upcoming](https://github.com/TomKellyGenetics/leiden/tree/dev) 0.4 release of "leiden". The purpose of this issue is to notify Seurat maintainers of upcoming changes to the backend of "leiden" in R (legacy behaviour will be retained). I will open a PR with the suggested changes if necessary when releasing the new version of "leiden".

I would rather not change default parameters (just compute them more efficiently) unless there's a clear benefit. However, there is limited functionality available in the `igraph::cluster_leiden()` and it does not support all of the cost functions so using "modularity" may be preferable. This change may be more consistent implementations of Louvain in `Suerat::FindClusters`.

**Comment by denvercal1234GitHub** [comment]
Thank you, Tom!

**Comment by Jay-Leung** [comment]
Hi @TomKellyGenetics , 

I have been using Seurat::FindClusters with Leiden and the performance is quite slow, especially if I am running various permutations to determine the resolution, params, and PCs to use for clustering. Is there any update on calling leiden in r-igraph with Seurat::FindClusters, now that igraph 1.2.7 is released? 

Thank you so much!

Regards,
Jay

**Comment by TomKellyGenetics** [comment]
Hi @Jay-Leung, thanks for chasing this up.

To clarify, I've been in contact with @vtraag (developer for python "leidenalg" and contributor to r-igraph) and have been expecting this for many months.

The development (branch "dev" on TomKellyGenetics/leiden) version of "leiden" in R already supports calling igraph where possible, including parameters called by Suerat. https://github.com/TomKellyGenetics/leiden/issues/1

No further changes to Suerat::FindClusters will be needed in my opinion. All that is needed is to merge the "dev" version of "leiden" with changes to master made since then to maintain the package (i.e., keep it on CRAN).

This should not take much time as I have prepared for the release of igraph but for full disclosure, I have family commitments and have taken a non-academic job a few months ago. So I do not have much time to maintain this package any more (but it is on my list!). Thanks for understanding. 

In the meantime, if you have the updated version of igraph installed, you are welcome to try installing the updated package from GitHub.

    devtools::install_github("TomKellyGenetics/leiden", ref = "dev")

**Comment by Jay-Leung** [comment]
Hi @TomKellyGenetics ,

Thanks for the prompt reply despite your hectic schedule! I will try out what you suggested. However, I ran into an error trying to install the development version of leiden. I have pasted the output of the console below. I had no issues installing the current version of leiden, so I am not sure what is wrong. I have also updated igraph and reticulate to versions 1.2.7 and 1.22 respectively. Thank you!

`> devtools::install_github("TomKellyGenetics/leiden", ref = "dev")`
Downloading GitHub repo TomKellyGenetics/leiden@dev
√  checking for file 'C:\Users\User\AppData\Local\Temp\RtmpUl9vge\remotes2bc44b966de7\TomKellyGenetics-leiden-cba1a18/DESCRIPTION' (336ms)
-  preparing 'leiden': (377ms)
√  checking DESCRIPTION meta-information ... 
-  checking for LF line-endings in source and make files and shell scripts
-  checking for empty or unneeded directories
   Omitted 'LazyData' from DESCRIPTION
-  building 'leiden_0.4.1.9001.tar.gz'
   
Installing package into ‘C:/Users/User/Documents/R/win-library/4.1’
(as ‘lib’ is unspecified)
* installing *source* package 'leiden' ...
** using staged installation
** R
Error in parse(outFile) : 
  C:/Users/User/AppData/Local/Temp/Rtmp2NQij5/R.INSTALLe0019115cc2/leiden/R/find_partition.R:88:1: unexpected symbol
87: legacy = FALSE
88: degree_as_node_size
    ^
ERROR: unable to collate and parse R files for package 'leiden'
* removing 'C:/Users/User/Documents/R/win-library/4.1/leiden'
* restoring previous 'C:/Users/User/Documents/R/win-library/4.1/leiden'
Warning message:
In i.p(...) :
  installation of package ‘C:/Users/User/AppData/Local/Temp/RtmpUl9vge/file2bc43e613b88/leiden_0.4.1.9001.tar.gz’ had non-zero exit status

**Comment by TomKellyGenetics** [comment]
Thanks for reporting the bug @Jay-Leung. In this case it is a very simple fix as it is a syntax error. Sorry I don’t have time to test the updated release yet. Please bear in mind that several minor releases have been pushed to CRAN since I added the igraph leiden calls so some features may be missing from the development version.

**Comment by Jay-Leung** [comment]
Hi @TomKellyGenetics , I ran the installation again, and there is a repeated argument at line 88. Thanks for rectifying the bug so quickly. I understand the implications of using the development version, thanks a ton for helping with this!

`> devtools::install_github("TomKellyGenetics/leiden", ref = "dev")`
Downloading GitHub repo TomKellyGenetics/leiden@dev
√  checking for file 'C:\Users\User\AppData\Local\Temp\RtmpIjefAv\remotes24783bd25ed2\TomKellyGenetics-leiden-76ae5c3/DESCRIPTION' (397ms)
-  preparing 'leiden':
√  checking DESCRIPTION meta-information ... 
-  checking for LF line-endings in source and make files and shell scripts
-  checking for empty or unneeded directories
   Omitted 'LazyData' from DESCRIPTION
-  building 'leiden_0.4.1.9001.tar.gz'
   
Installing package into ‘C:/Users/User/Documents/R/win-library/4.1’
(as ‘lib’ is unspecified)
* installing *source* package 'leiden' ...
** using staged installation
** R
Error in parse(outFile) : 
  repeated formal argument 'degree_as_node_size' on line 88
ERROR: unable to collate and parse R files for package 'leiden'
* removing 'C:/Users/User/Documents/R/win-library/4.1/leiden'
Warning message:
In i.p(...) :
  installation of package ‘C:/Users/User/AppData/Local/Temp/RtmpIjefAv/file24785cb962ed/leiden_0.4.1.9001.tar.gz’ had non-zero exit status

**Comment by TomKellyGenetics** [comment]
This issue is now resolved with the release of leiden version 0.4.2 on CRAN.
https://github.com/TomKellyGenetics/leiden/releases/tag/0.4.2

https://cran.r-project.org/src/contrib/Archive/leiden/leiden_0.4.2.tar.gz
https://cran.r-project.org/web/packages/leiden/news/news.html

**Comment by yang050036** [comment]
Hi @TomKellyGenetics ,

Thanks for your great sofrware. I saw the code of Seurat::FindClusters to find that the leiden is still calls python with reticulate 
 in the annotation. And it is slowly when I used Seurat::FindClusters with `algorithm = 4, method = "igraph"` to cluster the cells. what we can do  to Accelerate my process？

**Comment by yang050036** [comment]
OK，I accelerate my process with `plan(multicore, workers = 11)`.

**Comment by TomKellyGenetics** [comment]
Thanks for your feedback, I’m pleased to hear it is compatible with multithreaded parameters. Clustering is an N^2 problem so it can be very time-consuming for larger numbers of cells. Note the R interface is provided for R user convenience but the Python and C implementations should have better performance on larger datasets.

If you have any feedback to improve the R wrapper called by Seurat, please file an issue or PR on the dedicated package for further discussion. Improvements to the documentation to describe multithreading settings is welcome too.
https://github.com/TomKellyGenetics/leiden
